### PR TITLE
Improve performance of FormattingHelper.CountDigits(uint)

### PR DESCRIPTION
**Pull request by stephentoub**
Branch-free, table-based lookup implementation.

|               Method |         Toolchain |      value |       Mean | Ratio | Code Size |
|--------------------- |------------------ |----------- |-----------:|------:|----------:|
|   AllLengthsToString | \main\corerun.exe |          ? | 132.378 ns |  1.00 |     315 B |
|   AllLengthsToString |   \pr\corerun.exe |          ? | 115.465 ns |  0.86 |     245 B |
|                      |                   |            |            |       |           |
|  AllLengthsTryFormat | \main\corerun.exe |          ? |  81.336 ns |  1.00 |     338 B |
|  AllLengthsTryFormat |   \pr\corerun.exe |          ? |  66.679 ns |  0.82 |     276 B |
|                      |                   |            |            |       |           |
|  FixedLengthToString | \main\corerun.exe |          1 |   1.278 ns |  1.00 |     278 B |
|  FixedLengthToString |   \pr\corerun.exe |          1 |   1.118 ns |  0.87 |     208 B |
|                      |                   |            |            |       |           |
| FixedLengthTryFormat | \main\corerun.exe |          1 |   2.948 ns |  1.00 |     272 B |
| FixedLengthTryFormat |   \pr\corerun.exe |          1 |   1.802 ns |  0.61 |     206 B |
|                      |                   |            |            |       |           |
|  FixedLengthToString | \main\corerun.exe |         12 |   9.863 ns |  1.00 |     278 B |
|  FixedLengthToString |   \pr\corerun.exe |         12 |   9.311 ns |  0.94 |     208 B |
|                      |                   |            |            |       |           |
| FixedLengthTryFormat | \main\corerun.exe |         12 |   4.427 ns |  1.00 |     272 B |
| FixedLengthTryFormat |   \pr\corerun.exe |         12 |   2.763 ns |  0.62 |     206 B |
|                      |                   |            |            |       |           |
|  FixedLengthToString | \main\corerun.exe |        123 |  11.306 ns |  1.00 |     278 B |
|  FixedLengthToString |   \pr\corerun.exe |        123 |   9.830 ns |  0.87 |     208 B |
|                      |                   |            |            |       |           |
| FixedLengthTryFormat | \main\corerun.exe |        123 |   6.108 ns |  1.00 |     272 B |
| FixedLengthTryFormat |   \pr\corerun.exe |        123 |   3.418 ns |  0.56 |     206 B |
|                      |                   |            |            |       |           |
|  FixedLengthToString | \main\corerun.exe |       1234 |  12.235 ns |  1.00 |     278 B |
|  FixedLengthToString |   \pr\corerun.exe |       1234 |  10.797 ns |  0.88 |     208 B |
|                      |                   |            |            |       |           |
| FixedLengthTryFormat | \main\corerun.exe |       1234 |   8.026 ns |  1.00 |     272 B |
| FixedLengthTryFormat |   \pr\corerun.exe |       1234 |   4.153 ns |  0.52 |     206 B |
|                      |                   |            |            |       |           |
|  FixedLengthToString | \main\corerun.exe |      12345 |  13.264 ns |  1.00 |     278 B |
|  FixedLengthToString |   \pr\corerun.exe |      12345 |  11.592 ns |  0.87 |     208 B |
|                      |                   |            |            |       |           |
| FixedLengthTryFormat | \main\corerun.exe |      12345 |   9.675 ns |  1.00 |     272 B |
| FixedLengthTryFormat |   \pr\corerun.exe |      12345 |   4.948 ns |  0.51 |     206 B |
|                      |                   |            |            |       |           |
|  FixedLengthToString | \main\corerun.exe |     123456 |  13.974 ns |  1.00 |     278 B |
|  FixedLengthToString |   \pr\corerun.exe |     123456 |  13.332 ns |  0.95 |     208 B |
|                      |                   |            |            |       |           |
| FixedLengthTryFormat | \main\corerun.exe |     123456 |  10.444 ns |  1.00 |     272 B |
| FixedLengthTryFormat |   \pr\corerun.exe |     123456 |   5.970 ns |  0.57 |     206 B |
|                      |                   |            |            |       |           |
|  FixedLengthToString | \main\corerun.exe |    1234567 |  15.367 ns |  1.00 |     278 B |
|  FixedLengthToString |   \pr\corerun.exe |    1234567 |  14.046 ns |  0.91 |     208 B |
|                      |                   |            |            |       |           |
| FixedLengthTryFormat | \main\corerun.exe |    1234567 |  12.668 ns |  1.00 |     272 B |
| FixedLengthTryFormat |   \pr\corerun.exe |    1234567 |   6.599 ns |  0.52 |     206 B |
|                      |                   |            |            |       |           |
|  FixedLengthToString | \main\corerun.exe |   12345678 |  16.666 ns |  1.00 |     278 B |
|  FixedLengthToString |   \pr\corerun.exe |   12345678 |  14.880 ns |  0.89 |     208 B |
|                      |                   |            |            |       |           |
| FixedLengthTryFormat | \main\corerun.exe |   12345678 |  14.293 ns |  1.00 |     272 B |
| FixedLengthTryFormat |   \pr\corerun.exe |   12345678 |   7.512 ns |  0.53 |     206 B |
|                      |                   |            |            |       |           |
|  FixedLengthToString | \main\corerun.exe |  123456789 |  18.165 ns |  1.00 |     278 B |
|  FixedLengthToString |   \pr\corerun.exe |  123456789 |  15.920 ns |  0.87 |     208 B |
|                      |                   |            |            |       |           |
| FixedLengthTryFormat | \main\corerun.exe |  123456789 |  16.049 ns |  1.00 |     272 B |
| FixedLengthTryFormat |   \pr\corerun.exe |  123456789 |   8.479 ns |  0.53 |     206 B |
|                      |                   |            |            |       |           |
|  FixedLengthToString | \main\corerun.exe | 1234567890 |  19.486 ns |  1.00 |     278 B |
|  FixedLengthToString |   \pr\corerun.exe | 1234567890 |  17.430 ns |  0.90 |     208 B |
|                      |                   |            |            |       |           |
| FixedLengthTryFormat | \main\corerun.exe | 1234567890 |  17.680 ns |  1.00 |     272 B |
| FixedLengthTryFormat |   \pr\corerun.exe | 1234567890 |   9.488 ns |  0.51 |     206 B |

```C#
private uint[] _values = new uint[] { 1, 1234567890, 12, 123456789, 123, 12345678, 1234, 1234567, 12345, 123456 };
private char[] _scratch = new char[20];

[Benchmark]
[Arguments(1)]
[Arguments(12)]
[Arguments(123)]
[Arguments(1234)]
[Arguments(12345)]
[Arguments(123456)]
[Arguments(1234567)]
[Arguments(12345678)]
[Arguments(123456789)]
[Arguments(1234567890)]
public string FixedLengthToString(uint value) => value.ToString();

[Benchmark]
[Arguments(1)]
[Arguments(12)]
[Arguments(123)]
[Arguments(1234)]
[Arguments(12345)]
[Arguments(123456)]
[Arguments(1234567)]
[Arguments(12345678)]
[Arguments(123456789)]
[Arguments(1234567890)]
public bool FixedLengthTryFormat(uint value) => value.TryFormat(_scratch, out _);

[Benchmark]
public int AllLengthsToString()
{
    int sum = 0;
    foreach (uint value in _values)
    {
        sum += value.ToString().Length;
    }
    return sum;
}

[Benchmark]
public bool AllLengthsTryFormat()
{
    bool success = true;
    foreach (uint value in _values)
    {
        success |= value.TryFormat(_scratch, out _);
    }
    return success;
}
```

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/FormattingHelpers.CountDigits.cs [hunk 1]
```diff
@@ -109,36 +111,52 @@ public static int CountDigits(ulong value)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static int CountDigits(uint value)
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/Text/FormattingHelpers.CountDigits.cs:112]: Update `CountDigits(ulong value)` above to take advantage of this?

**Review by tannergooding — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/FormattingHelpers.CountDigits.cs (hunk 1 above)
Comment (mattzink) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/Text/FormattingHelpers.CountDigits.cs:112]: > Update `CountDigits(ulong value)` above to take advantage of this?

FYI: There is a 64-bit version in the comments of Lemire's post

**Comment by stephentoub** [comment]
I'll merge this and follow-up on the 64-bit case.

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/FormattingHelpers.CountDigits.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/Text/FormattingHelpers.CountDigits.cs:112]: > FYI: There is a 64-bit version in the comments of Lemire's post

Which specifically? The variants I'd seen were only good up to a smaller number of bits, like 57.

Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/Text/FormattingHelpers.CountDigits.cs:112]: I assume Jan meant
```csharp
if (longValue <= uint.MaxValue)
    return CountDigits(unchecked((uint)longValue));
```
?

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/Text/FormattingHelpers.CountDigits.cs:112]: More like:
```
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static int CountDigits(ulong value)
        {
            int digits;
            uint part;
            if (value >= 1_000_000_000)
            {
                if (part >= 1_000_000_000_000_000_000)
                {
                    part = (uint)(value / 1_000_000_000_000_000_000);
                    digits = 18;
                }
                else
                {
                    part = (uint)(value / 1_000_000_000);
                    digits = 9;
                }
            }
            else
            {
                part = (uint)value;
                digits = 0;
            }
            
            return digits + CountDigits(part);
        }
```

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/Text/FormattingHelpers.CountDigits.cs:112]: > More like

Yeah, that's what I was planning on doing, but I want to experiment.

**Pull request merged**